### ISSUE TITLE
Can Commands API have an Async endpoint as well?

### ISSUE THREAD

**Issue by blueelvis** [projectkudu/kudu#2828]
### Project structures.
Drupal/Wordpress (Running on Web App for Containers)

### Mention any other details that might be useful.
Is it possible to have an Async endpoint for the `POST /api/command` route as well like we have for Zip Deploy? And maybe even have webhook capability?

An example would unzipping the production website's uploaded content to dev & staging. We have a large zip file which times out in some cases even when using the Command api (Zip deploy always times out).

What do you guys think?

Thanks,
Pranav

**Comment by davidebbo** [comment]
I think it would indeed be nice to have, but frankly not supper likely to happen since there hasn't been much request for this.

When you say "Zip deploy always times out", do you mean even with the async endpoint?

**Comment by blueelvis** [comment]
@davidebbo  -- That looks interesting. But, what if I want to deploy a ZIP file to a particular folder? For example, `/home/site/wwwroot/sites/default/files`? Does, the Async ZIP Deploy api support this?

Is there any kind of timeout or size limitation associated with the Async Zip deploy feature as well?

**Comment by davidebbo** [comment]
@blueelvis the zipdeploy API currently only supports deploying to the wwwroot folder.

The `/api/zip` API supports unzipping a file in any folder, but it is not really a deployment API (see [here](https://blogs.msdn.microsoft.com/appserviceteam/2017/10/16/zip-push-deployment-for-web-apps-functions-and-webjobs/) for some discussion on that). Also, it has no async flavor.

**Comment by blueelvis** [comment]
@davidebbo - Looks neat. The article says that a response will be returned as soon as the ZIP file is uploaded. But, if I am correct, Azure has a timeout of 230 seconds so if I assume an upload speed of 50 MBps, I should be able to upload a zip file of around 1150 MB correct?

Also, if the ZIP has got nested folder structure, will that be maintained? For example, even if the API deploys just to `wwwroot`, I can have a zip containing `sites/default/files/settings.php` and it would get properly deployed while maintaining the folder structure inside `wwwroot`?

**Comment by davidebbo** [comment]
230 seconds is an **idle timeout**, so should not apply during an upload where bits are flowing.

The zipdeploy API preserves the folder that's in the zip. But not that if you call it multiple times, it will remove any files that were in the previous zip and are not in the current one.

**Comment by blueelvis** [comment]
@davidebbo -- Hmm I didn't know that it was an `idle timeout` . So, if a long running command is producing output, things should work as expected right?

Asking because we are dealing with timeout issues for an unzip. Kudu seems to gather all the output at the end of command exit and then write it back. Cannot be it something like a `git push` where bits are flowing both sides?

**Comment by davidebbo** [comment]
@blueelvis to further clarify, I'm referring to an **idle connection timeout**, which means data has to be flowing in either the request or response. If you just start a really long command via Command API, that is not the case because nothing gets sent to the response until it is complete.

**Comment by blueelvis** [comment]
Uh huh. Thanks @davidebbo for the explanation. Do you have anything in mind to run a command in an async way though? Is there support coming to Linux for Webjobs?

**Comment by davidebbo** [comment]
@blueelvis I think it comes back to the original question about adding some async flavor of the Commands API. For this to happen, someone would need to pick it up :)

Not an expert on the Linux side, but as far as I know there is no plan to bring WebJobs there.

**Comment by blueelvis** [comment]
@davidebbo - Hehe. Someone needs to definitely pick this up. But, other than that, could you please explain the differences between the Kudu which is deployed for Windows Web Apps & Kudu which is deployed on Linux Web Apps?

Asking because I have been wondering if there is a way to deploy privately built Kudu. It seems as if Kudu which is available here and which is actually deployed on Azure are slightly different. Please feel free to correct me on this though.

Issue related to privately built Kudu - https://github.com/projectkudu/kudu/issues/2815

Thanks!

-Pranav

**Comment by davidebbo** [comment]
It's the same Kudu binaries, but there is a lot of conditional logic that makes the Linux behavior different (you can see that in the sources). Only a subset of features is supported on Linux (e.g. no WebJobs).

**Comment by stevengum** [comment]
> ### Mention any other details that might be useful.
> Is it possible to have an Async endpoint for the `POST /api/command` route as well like we have for Zip Deploy? And maybe even have webhook capability?
> 
> An example would unzipping the production website's uploaded content to dev & staging. We have a large zip file which times out in some cases even when using the Command api (Zip deploy always times out).

I don't have the same scenario as @blueelvis but rather when calling `npm install` the client performing the request can timeout while waiting for the install to finish. In testing, adding the [`--no-progress`](https://github.com/npm/npm/issues/11283) flag doesn't change the performance noticeably (or actually made it worse) on the latest npm.

@davidebbo you indicated in [another thread](https://github.com/projectkudu/kudu/issues/2474#issuecomment-307404157) the following:

> AWahlqvist the **230s timeout is an App Service limit which unfortunately we have no control over.** The one way to avoid it is to keep some data flowing. i.e. write to the response to keep it live. That's what we do in the 'git push' scenario. But it's harder to do in an API that needs to return a proper http code, as the minute you write the headers, you lose the ability to change the code.
> 
> The proper solution would be to provide an async pattern for this API, which is what you are suggesting. But that's a fair bit of work.

Is this 230s timeout limit still unchangeable by developers and by the Kudu team?

**Comment by ahmelsayed** [comment]
> Is this 230s timeout limit still unchangeable by developers and by the Kudu team?

Yes. It's a limit from networking on idle open connection with no traffic. I don't think that will change.
